File: legacy-python/bot/handlers/availability.py
"""Команда /availability — общие свободные окна (в том числе по кворуму)."""

from __future__ import annotations

import html
import logging
import re

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.constants import ChatType, ParseMode
from telegram.ext import CallbackQueryHandler, CommandHandler, ContextTypes

from qairu.core.availability import (
    AvailabilityResult,
    compute_availability,
    parity_from_semester_start,
)
from qairu.core.intervals import fmt_interval, fmt_minutes
from qairu.db import repo
from qairu.db.models import Chat, User
from qairu.i18n import format_day, t
from qairu.textutils import clip
from ..utils import (
    chat_tz,
    extract_mentioned_users,
    extract_min_duration,
    mention_list,
    plain_names,
    sync_user,
    today_in,
)

log = logging.getLogger(__name__)

PERCENT_RE = re.compile(r"^(\d{1,3})%$")
QUORUM_RE = re.compile(r"^(?:q|кворум|кв)(\d{1,3})$", re.IGNORECASE)


def parse_quorum(args: list[str], total: int) -> int | None:
    """`80%` → 80% участников, `q8` / `кворум8` → ровно 8 человек. None — нужны все."""
    for index, arg in enumerate(args):
        percent = PERCENT_RE.match(arg)
        if percent:
            value = int(percent.group(1))
            return max(1, min(total, round(total * value / 100)))
        absolute = QUORUM_RE.match(arg)
        if absolute:
            return max(1, min(total, int(absolute.group(1))))
        if arg.lower() in ("кворум", "quorum", "кв", "q") and index + 1 < len(args):
            nxt = args[index + 1]
            if nxt.isdigit():
                return max(1, min(total, int(nxt)))
    return None


async def cmd_availability(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = update.effective_message
    tg_chat = update.effective_chat
    tg_user = update.effective_user
    if not message or not tg_chat or not tg_user:
        return

    async with repo.session() as session:
        await sync_user(session, tg_user)
        await session.commit()

    if tg_chat.type in (ChatType.GROUP, ChatType.SUPERGROUP):
        async with repo.session() as session:
            await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
            await repo.add_membership(session, tg_chat.id, tg_user.id)
            await session.commit()
        text = await _build_report(update, context, tg_chat.id)
        await message.reply_text(text, parse_mode=ParseMode.HTML, disable_web_page_preview=True)
        return

    # личка: выбрать, про какой чат речь
    async with repo.session() as session:
        chats = await repo.user_chats(session, tg_user.id)
        user = await session.get(User, tg_user.id)
        lang = user.lang if user else "ru"

    if not chats:
        await message.reply_text(t(lang, "avail_no_members"))
        return
    if len(chats) == 1:
        text = await _build_report(update, context, chats[0].chat_id)
        await message.reply_text(text, parse_mode=ParseMode.HTML, disable_web_page_preview=True)
        return

    args = " ".join(context.args or [])[:48]
    keyboard = InlineKeyboardMarkup(
        [
            [InlineKeyboardButton(chat.title or str(chat.chat_id),
                                  callback_data=f"avl:{chat.chat_id}")]
            for chat in chats
        ]
    )
    context.user_data["avail_args"] = args
    await message.reply_text(t(lang, "avail_title", who="…"), parse_mode=ParseMode.HTML,
                             reply_markup=keyboard)


async def on_chat_choice(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    await query.answer()
    chat_id = int(query.data.split(":", 1)[1])
    saved = (context.user_data.get("avail_args") or "").split()
    text = await _build_report(update, context, chat_id, forced_args=saved)
    await query.edit_message_text(text, parse_mode=ParseMode.HTML, disable_web_page_preview=True)


async def _build_report(
    update: Update,
    context: ContextTypes.DEFAULT_TYPE,
    chat_id: int,
    forced_args: list[str] | None = None,
) -> str:
    tg_user = update.effective_user
    args = forced_args if forced_args is not None else (context.args or [])

    async with repo.session() as session:
        chat = await repo.get_chat(session, chat_id)
        if chat is None:
            return t("ru", "chat_not_setup")
        lang = chat.lang

        members = await repo.chat_members(session, chat_id)
        if not members:
            return t(lang, "avail_no_members")

        mentioned, unknown = await extract_mentioned_users(update, session)
        if mentioned:
            caller = await session.get(User, tg_user.id)
            selected, seen = [], set()
            for user in ([caller] if caller else []) + mentioned:
                if user and user.user_id not in seen:
                    seen.add(user.user_id)
                    selected.append(user)
            who = t(lang, "avail_selected", count=len(selected), names=plain_names(selected))
        else:
            selected = members
            who = t(lang, "avail_all", count=len(selected))

        minimum = extract_min_duration(args, chat.min_slot_min)
        people = await repo.build_person_schedules(session, selected)
        id_to_user = {user.user_id: user for user in selected}

    with_data = [person for person in people if person.has_data]
    if not with_data:
        return t(lang, "avail_need_schedule")

    quorum = parse_quorum(args, len(with_data))
    parity_of = parity_from_semester_start(chat.semester_start) if chat.semester_start else None
    tz = chat_tz(chat)
    today = today_in(tz)

    result = compute_availability(
        people,
        start_day=today,
        days_ahead=7,
        day_start=chat.day_start_min,
        day_end=chat.day_end_min,
        min_slot=minimum,
        quorum=quorum,
        parity_of=parity_of,
        buffer_min=chat.travel_buffer_min,
    )

    text = _render(result, lang, who, id_to_user, chat, parity_of, today)
    if unknown:
        text += "\n" + "\n".join(
            t(lang, "avail_user_not_found", name=html.escape(name)) for name in unknown[:3]
        )
    # Telegram режет сообщения длиннее 4096 символов — на большой группе
    # список окон легко перевалит лимит, и сообщение просто не уйдёт.
    return clip(text, note=t(lang, "avail_clipped"))


def _render(result: AvailabilityResult, lang: str, who: str, id_to_user: dict[int, User],
            chat: Chat, parity_of, today) -> str:
    total = len(result.participants)

    if result.everyone:
        text = t(lang, "avail_title", who=who)
    else:
        text = t(lang, "avail_quorum_title", quorum=result.quorum, total=total, who=who)

    if parity_of:
        parity_key = "parity_odd" if parity_of(today) == 0 else "parity_even"
        text += t(lang, "avail_parity_note", parity=t(lang, parity_key))

    if not result.any_windows:
        key = "avail_none" if result.everyone else "avail_quorum_none"
        text += "\n" + t(lang, key, min=result.min_slot, quorum=result.quorum)
    else:
        for day in result.days:
            if not day.has_windows:
                continue
            text += f"\n<b>{format_day(lang, day.day)}</b>\n"
            for window in day.windows:
                text += f"<code>{fmt_interval(window.interval)}</code>\n"
                if not result.everyone:
                    absent = [
                        id_to_user[person.user_id]
                        for person in result.participants
                        if person.user_id not in window.free_ids and person.user_id in id_to_user
                    ]
                    if absent:
                        text += t(lang, "avail_quorum_line", count=window.count, total=total,
                                  missing=plain_names(absent)) + "\n"
                    else:
                        text += t(lang, "avail_quorum_all") + "\n"

    if result.missing:
        missing_users = [id_to_user[p.user_id] for p in result.missing if p.user_id in id_to_user]
        text += t(lang, "avail_missing_note", names=mention_list(missing_users))

    text += t(lang, "avail_hint", start=fmt_minutes(chat.day_start_min),
              end=fmt_minutes(chat.day_end_min), min=result.min_slot)
    return text


def register(app) -> None:
    app.add_handler(CommandHandler("availability", cmd_availability))
    app.add_handler(CommandHandler("free", cmd_availability))
    app.add_handler(CallbackQueryHandler(on_chat_choice, pattern=r"^avl:"))

File: src/bot/handlers/availability.ts
/** Команда /free (она же /availability) — общие свободные окна в группе, в том числе по кворуму. */

import "server-only";

import {
  type AvailabilityResult,
  type ParityOf,
  anyWindows,
  computeAvailability,
  parityFromSemesterStart,
} from "@/core/availability";
import { fmtInterval, fmtMinutes } from "@/core/intervals";
import { clip, escapeHtml } from "@/core/textutils";
import { type DateStr, chatTz, todayIn } from "@/core/timeutils";
import * as repo from "@/db/repo";
import type { Chat, User } from "@/db/schema";
import { formatDay, t } from "@/i18n";
import { isGroup, sendMessage, type TgMessage } from "../api";
import {
  extractMentionedUsers,
  extractMinDuration,
  mentionList,
  plainNames,
  syncUser,
} from "../context";

const PERCENT_RE = /^(\d{1,3})%$/;
const QUORUM_RE = /^(?:q|кворум|кв)(\d{1,3})$/i;

/** `80%` → 80% участников, `q8` / `кворум8` → ровно 8 человек. null — нужны все. */
export function parseQuorum(args: string[], total: number): number | null {
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    const percent = PERCENT_RE.exec(arg);
    if (percent) {
      const value = Number(percent[1]);
      return Math.max(1, Math.min(total, Math.round((total * value) / 100)));
    }
    const absolute = QUORUM_RE.exec(arg);
    if (absolute) {
      return Math.max(1, Math.min(total, Number(absolute[1])));
    }
    if (["кворум", "quorum", "кв", "q"].includes(arg.toLowerCase()) && index + 1 < args.length) {
      const next = args[index + 1];
      if (/^\d+$/.test(next)) return Math.max(1, Math.min(total, Number(next)));
    }
  }
  return null;
}

/** /free в группе: общие окна участников этого чата. В личке команда ведёт на сайт (см. router). */
export async function cmdAvailability(message: TgMessage, args: string[]): Promise<void> {
  const from = message.from;
  if (!from || !isGroup(message.chat)) return;

  await syncUser(from);
  await repo.upsertChat(message.chat.id, message.chat.title ?? "");
  await repo.addMembership(message.chat.id, from.id);
  const text = await buildReport(message, message.chat.id, args);
  await sendMessage({
    chat_id: message.chat.id,
    text,
    parse_mode: "HTML",
    link_preview_options: { is_disabled: true },
  });
}

async function buildReport(
  message: TgMessage,
  chatId: number,
  args: string[],
): Promise<string> {
  const chat = await repo.getChat(chatId);
  if (!chat) return t("ru", "chat_not_setup");
  const lang = chat.lang;

  const members = await repo.chatMembers(chatId);
  if (members.length === 0) return t(lang, "avail_no_members");

  const { found: mentioned, missing: unknown } = await extractMentionedUsers(message);

  let selected: User[];
  let who: string;
  if (mentioned.length) {
    const caller = message.from ? await repo.getUser(message.from.id) : null;
    const seen = new Set<number>();
    selected = [];
    for (const user of [...(caller ? [caller] : []), ...mentioned]) {
      if (!seen.has(user.userId)) {
        seen.add(user.userId);
        selected.push(user);
      }
    }
    who = t(lang, "avail_selected", { count: selected.length, names: plainNames(selected) });
  } else {
    selected = members;
    who = t(lang, "avail_all", { count: selected.length });
  }

  const minimum = extractMinDuration(args, chat.minSlotMin);
  const people = await repo.buildPersonSchedules(selected);
  const idToUser = new Map(selected.map((user) => [user.userId, user]));

  const withData = people.filter((person) => person.hasData);
  if (withData.length === 0) return t(lang, "avail_need_schedule");

  const quorum = parseQuorum(args, withData.length);
  const parityOf = chat.semesterStart ? parityFromSemesterStart(chat.semesterStart) : null;
  const today = todayIn(chatTz(chat));

  const result = computeAvailability(people, today, {
    daysAhead: 7,
    dayStart: chat.dayStartMin,
    dayEnd: chat.dayEndMin,
    minSlot: minimum,
    quorum,
    parityOf,
    bufferMin: chat.travelBufferMin,
  });

  let text = render(result, lang, who, idToUser, chat, parityOf, today);
  if (unknown.length) {
    text +=
      "\n" +
      unknown
        .slice(0, 3)
        .map((name) => t(lang, "avail_user_not_found", { name: escapeHtml(name) }))
        .join("\n");
  }
  // Telegram режет сообщения длиннее 4096 символов — на большой группе
  // список окон легко перевалит лимит, и сообщение просто не уйдёт.
  return clip(text, 3900, t(lang, "avail_clipped"));
}

function render(
  result: AvailabilityResult,
  lang: string,
  who: string,
  idToUser: Map<number, User>,
  chat: Chat,
  parityOf: ParityOf | null,
  today: DateStr,
): string {
  const total = result.participants.length;

  let text = result.everyone
    ? t(lang, "avail_title", { who })
    : t(lang, "avail_quorum_title", { quorum: result.quorum, total, who });

  if (parityOf) {
    const parityKey = parityOf(today) === 0 ? "parity_odd" : "parity_even";
    text += t(lang, "avail_parity_note", { parity: t(lang, parityKey) });
  }

  if (!anyWindows(result)) {
    const key = result.everyone ? "avail_none" : "avail_quorum_none";
    text += "\n" + t(lang, key, { min: result.minSlot, quorum: result.quorum });
  } else {
    for (const day of result.days) {
      if (day.windows.length === 0) continue;
      text += `\n<b>${formatDay(lang, day.day)}</b>\n`;
      for (const window of day.windows) {
        text += `<code>${fmtInterval(window.interval)}</code>\n`;
        if (result.everyone) continue;
        const absent = result.participants
          .filter(
            (person) => !window.freeIds.includes(person.userId) && idToUser.has(person.userId),
          )
          .map((person) => idToUser.get(person.userId)!);
        text += absent.length
          ? t(lang, "avail_quorum_line", {
              count: window.freeIds.length,
              total,
              missing: plainNames(absent),
            }) + "\n"
          : t(lang, "avail_quorum_all") + "\n";
      }
    }
  }

  if (result.missing.length) {
    const missingUsers = result.missing
      .filter((person) => idToUser.has(person.userId))
      .map((person) => idToUser.get(person.userId)!);
    text += t(lang, "avail_missing_note", { names: mentionList(missingUsers) });
  }

  text += t(lang, "avail_hint", {
    start: fmtMinutes(chat.dayStartMin),
    end: fmtMinutes(chat.dayEndMin),
    min: result.minSlot,
  });
  return text;
}
